File: Data Structures & Algorithms/three-integer-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int l = i + 1;
            int r = n - 1;
            while (l < r) {
                int curSum = nums[i] + nums[l] + nums[r];
                if (curSum == 0) {
                    res.push_back({nums[i], nums[l], nums[r]});
                    l += 1;
                    r -= 1;
                    while (l < r && nums[l] == nums[l - 1]) {
                        l += 1;
                    }
                } else if (curSum < 0) {
                    l += 1;
                } else {
                    r -= 1;
                }
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-1.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        """
        Input: nums = [2,-1,1,2], k = 2




        """
        res = 0
        prefix_map = defaultdict(int)
        prefix_map[0] = 1
        prefix_sum = 0

        for num in nums:
            prefix_sum += num

            if (prefix_sum - k) in prefix_map:
                res += prefix_map[prefix_sum - k]
            
            prefix_map[prefix_sum] += 1
        
        return res